ldp/graph/common_ops.py

```diff
@@ -283,6 +297,35 @@ async def forward(
 
         return result.messages[0]
 
+    async def _call_single_and_maybe_validate(
+        self, model: LLMModel, num_retries: int, **kwargs
+    ) -> LLMResult:
+        if not self.response_validator:
+            # If a response validator is not supplied, then we should not do any retries here - leave
+            # that for LiteLLM to handle.
+            return await model.call_single(**kwargs)
+
+        @tenacity.retry(
+            retry=tenacity.retry_if_exception_type(ResponseValidationError),
+            # num_retries+1 because the first call is not a retry
+            stop=tenacity.stop_after_attempt(num_retries + 1),
+            wait=tenacity.wait_fixed(1),
+        )
+        async def call_and_validate() -> LLMResult:
+            result = await model.call_single(**kwargs)
+
+            validated = cast(Callable, self.response_validator)(result)
+            if is_coroutine_callable(self.response_validator):
+                validated = await validated
+            if not validated:
+                # Don't provide more info here - user can control error logging
+                # in the validator if desired.
+                raise ResponseValidationError
```

Comment:
```suggestion
                raise ResponseValidationError(f"Validator {self.response_validator} failed.)
```

Let's put a message here because in log files what happens is this shows up as blank line, unless your log formatter also shows the exception type

Reply:
Alright, when doing this I saw the light - it's nice to be able to chain exceptions and see the full traceback (RetryError -> ResponseValidationError -> original exception). I've switched to the pattern you suggested here: https://github.com/Future-House/ldp/pull/212#discussion_r1917863172